docs: describe OIDC for native apps (#1379)
ory/docs: docs/kratos/social-signin/96_native-apps.mdx

---
id: native-apps

> title: Social sign-in for native apps

Reviewer: ```suggestion
title: Social sign-in for native and mobile apps
```

sidebar_label: Native apps
---

```mdx-code-block
import Mermaid from '@theme/Mermaid';
import CodeFromRemote from "@theme/CodeFromRemote"
```

# Social sign-in for native applications

## Overview

> This page covers how to implement social sign-in for native applications via OIDC and OAUth 2.0. The user interaction looks like

Reviewer: ```suggestion
This page covers how to implement social sign-in for native applications via OIDC and OAuth 2.0. The user interaction looks like
```

this:

- The user is presented with a login or registration screen that includes a social sign-in button.
- The user clicks the social sign-in button. A browser window opens (using
  [ASWebAuthenticationSession](https://developer.apple.com/documentation/authenticationservices/aswebauthenticationsession) on iOS
  or [Custom Tabs](https://developer.chrome.com/docs/android/custom-tabs/) on Android).
- The user authenticates with the identity provider and grants the application access to profile information.
- The user is redirected back to the application and is logged in.

## The native app authentication flow

> From a high level, the native app initializes a login or registration flow and receives the first part of the session token
> exchange code from the Ory Network. After the user performed the social sign-in, the user is redirected back to the native

Reviewer: ```suggestion
From a high level, the native app initializes a login or registration flow and receives the first part of the Session Token
Exchange Code from the Ory Network. After the user performed the social sign-in, the user is redirected back to the native
```

Capitalize because it is an Ory Network specific term?

Do we need to define Session Token Exchange Code as an Ory Network specific term here, or is it clear?

Author: I'd not capitalize it as it is not a brand name but just some term.

application via an [iOS Universal Link](https://developer.apple.com/ios/universal-links/) or

> [Android App Link](https://developer.android.com/training/app-links). The native application then exchanges the session token
> exchange code for a session token.

Reviewer: ```suggestion
[Android App Link](https://developer.android.com/training/app-links). The native application then exchanges the Session Token
Exchange Code for a Session Token.
```

The flow looks like this:

```mdx-code-block
<Mermaid chart={`
sequenceDiagram
  actor U as user
  participant A as Native App
  participant B as Browser
  participant O as Ory Network
  participant I as Identity Provider
  autonumber
  U->>+A: Open login screen
  A->>+O: create API flow with return_session_token_exchange_code=true
  O->>A: flow.session_token_exchange_code: ...
  U->>A: choose SSO provider
  A->>O: submit form with provider=...
  O->>A: Status 422 with err.response.data.redirect_browser_to=<IdP auth URL>
  A->>+B: go to <IdP auth URL>
  B->>+I: open <IdP auth URL>
  U->>I: login, consent
  I->>-B: redirect to <IdP callback URL>
  B->>O: open <IdP callback URL>
  deactivate B
  O->>A: redirect with App link to <return to>/?code=...
  A->>O: exchange code for session token
  O->>-A: session token
`}/>
```

### Implementation

The following sections describe how to implement the native app authentication flow. The code examples are written in TypeScript

> for React Native. The numbers in the headlines refer to the steps in the flow above.

Reviewer: ```suggestion
for React Native. The steps refer to the steps in the flow diagram above.
```

#### Steps 1-5: Create an API flow with return_session_token_exchange_code=true

> The user opens the login screen (1) and the native application initializes a login or registration flow through the Ory Network

Reviewer: What do you think about using a layout like:

1. The user opens the login screen 
2. The native application initializes a login or registration flow through the Ory Network APIs with the following parameters:

IMO this would make it easier to parse and map the steps to the diagram above. Also happy to take a whack at it if you agree.

  (3).

<CodeFromRemote
  lang="ts"
  src="https://github.com/ory/kratos-selfservice-ui-react-native/blob/2ef61d58aac508ad8890c01dac4b5fd4740265ff/src/components/Routes/Login.tsx"
  startAt="  const initializeFlow = () =>"
  endAt="      .then(({ data: f }) => setFlow(f))"
/>

Upon rendering the form, the user selects the specific social sign-in provider (4). The native application submits the form to the
Ory Network (5).

#### Steps 6-12: Open the identity provider authentication URL in a browser

Ory Network returns a `422` status code if the user needs to perform a social sign-in from a flow of type `api` (6). The response
contains the URL of the identity provider in the `redirect_browser_to` field.

<CodeFromRemote
  lang="ts"
  src="https://github.com/ory/kratos-selfservice-ui-react-native/blob/2ef61d58aac508ad8890c01dac4b5fd4740265ff/src/helpers/form.tsx"
  startAt="        case 422:"
  endAt="          )"
/>

[WebBrowser.openAuthSessionAsync](https://docs.expo.dev/versions/latest/sdk/webbrowser/#webbrowseropenauthsessionasyncurl-redirecturl-options)

> opens a browser tab for authentication for the specified `url` (7+8). Impementations in other languages and frameworks may vary.

Reviewer: ```suggestion
opens a browser tab for authentication for the specified `url` (7+8). Implementations in other languages and frameworks may vary.
```

  startAt="  const result = await WebBrowser.openAuthSessionAsync("
  endAt="  )"
/>

The user authenticates with the identity provider and grants the application access to profile information (9). The identity
provider then issues a redirect to the Ory Network callback URL (10), which the browser follows (11).

#### Steps 12-14: Exchange the session token exchange code for a session token

Finally, Ory Network issues a session for the user and redirects the browser to the application's `return_to` URL (12). The native
application receives the second part of the session token exchange code in the `code` URL query parameter. In the example below,
the call to
[WebBrowser.maybeCompleteAuthSession](https://docs.expo.dev/versions/latest/sdk/webbrowser/#webbrowsermaybecompleteauthsessionoptions)
is used to close the browser tab and return the control flow to the code that awaited the call to
`WebBrowser.openAuthSessionAsync`.

<CodeFromRemote
  lang="tsx"
  src="https://github.com/ory/kratos-selfservice-ui-react-native/blob/2ef61d58aac508ad8890c01dac4b5fd4740265ff/src/components/Routes/Callback.tsx"
  startAt="const Callback = (props: Props) => {"
  endAt="  )"
/>

> The native application then exchanges the session token exchange code for a session token (13) using the first part of the code

Reviewer: Capitalize Ory Network specific term?

returned from the flow initialization, and the second part of the code returned from the `return_to` query parameter.

<CodeFromRemote
  lang="ts"
  src="https://github.com/ory/kratos-selfservice-ui-react-native/blob/2ef61d58aac508ad8890c01dac4b5fd4740265ff/src/helpers/form.tsx"
  startAt="  const fetchToken = (params: FrontendApiExchangeSessionTokenRequest) =>"
  endAt='    newOrySdk("").exchangeSessionToken(params)'
/>

> Finally, the native application stores the session token in the secure storage (14) and redirects the user to the home screen.

Reviewer: Capitalize Ory Network specific term?
